[JIT] Resolve `torch.device` in recursive compilation of classes

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47734 [JIT] Resolve `torch.device` in recursive compilation of classes**

**Summary**
This commit allows `torch.device` to be resolved properly when used in
class types that are recursively scripted. This is accomplished by augmenting
the resolution callback used during recursively class scripting to include
the type annotations used on class method declarations.

Classes that are not explicitly annotated with `@torch.jit.script` are
implicitly scripted during the compilation of a function or class method
that uses them. One key difference between this method of class type
compilation and explicit scripting is that the former uses a resolution callback
that can only resolve variables that class methods close over (see
`_jit_internal.createResolutionCallbackForClassMethods`). This does
not include type annotations and default arguments. This means that
builtin types like `torch.Tensor` and `torch.device` cannot be resolved
using the resolution callback. This issue does not arise when explicitly
scripting classes because the resolution callback for that code path is
constructed from scope of the class definition
(see `_jit_internal.createResolutionCallbackFromFrame`). `torch.Tensor`
and `torch.device` are almost always present in that scope, usually from
`import`ing `torch`.

**Test Plan**
This commit adds a new unit test to `TestClassType`,
`test_recursive_script_builtin_type_resolution`.

**Fixes**
This commit closes #47405.

Differential Revision: [D24995374](https://our.internmc.facebook.com/intern/diff/D24995374)

Commit:
[JIT] Resolve `torch.device` in recursive compilation of classes

**Summary**
This commit allows `torch.device` to be resolved properly when used in
class types that are recursively scripted. This is accomplished using
`string_to_type_lut`, similar to how `torch.Tensor` is resolved in this
situation.

Classes that are not explicitly annotated with `@torch.jit.script` are
implicitly scripted during the compilation of a function or class method
that uses them. One key difference between this method of class type
compilation and explicit scripting is that the former uses a resolution callback
that can only resolve variables that class methods close over (see
`_jit_internal.createResolutionCallbackForClassMethods`). This does
not include type annotations and default arguments. This means that
builtin types like `torch.Tensor` and `torch.device` cannot be resolved
using the resolution callback. This issue does not arise when explicitly
scripting classes because the resolution callback for that code path is
constructed from scope of the class definition
(see `_jit_internal.createResolutionCallbackFromFrame`). `torch.Tensor`
and `torch.device` are almost always present in that scope, usually from
`import`ing `torch`.

**Test Plan**
This commit adds a new unit test to `TestClassType`,
`test_recursive_script_builtin_type_resolution`.

**Fixes**
This commit closes #47405.

[ghstack-poisoned]

Commit:
Update on "[JIT] Resolve `torch.device` in recursive compilation of classes"


**Summary**
This commit allows `torch.device` to be resolved properly when used in
class types that are recursively scripted. This is accomplished using
`string_to_type_lut`, similar to how `torch.Tensor` is resolved in this
situation.

Classes that are not explicitly annotated with `@torch.jit.script` are
implicitly scripted during the compilation of a function or class method
that uses them. One key difference between this method of class type
compilation and explicit scripting is that the former uses a resolution callback
that can only resolve variables that class methods close over (see
`_jit_internal.createResolutionCallbackForClassMethods`). This does
not include type annotations and default arguments. This means that
builtin types like `torch.Tensor` and `torch.device` cannot be resolved
using the resolution callback. This issue does not arise when explicitly
scripting classes because the resolution callback for that code path is
constructed from scope of the class definition
(see `_jit_internal.createResolutionCallbackFromFrame`). `torch.Tensor`
and `torch.device` are almost always present in that scope, usually from
`import`ing `torch`.

**Test Plan**
This commit adds a new unit test to `TestClassType`,
`test_recursive_script_builtin_type_resolution`.

**Fixes**
This commit closes #47405.

[ghstack-poisoned]

eellison reviewed (commented):
This will make `torch.device` resolve, but i think we should try to fix the underlying problem if possible. There's nothing that's preventing users from using a different type alias for `torch.device`, and we ran into all sorts of user bugs previously when we didn't use the actual python object to resolve the type.

SplitInfinity commented:
I think the best solution is to fix the resolution callback and use the same one that would be used with explicit scripting (i.e. one constructed using `createResolutionCallbackFromFrame` instead of `createResolutionCallbackForClassMethods`). Such a callback would include `torch.device`, the Python object, as well as any aliases for it because it would incorporate all locals and globals from the scope in which the class is defined. 

But how can that be done without knowing which scope to use to create this aforementioned resolution callback?

eellison commented:
```
import torch
from torch import device

class A:
    def __init__(self):
        pass

    def f(self, x: torch.Tensor, y: device):
        return x.to(device=y)

import typing

print(typing.get_type_hints(A.f))
# {'x': <class 'torch.Tensor'>, 'y': <class 'torch.device'>}
```

I haven't looked into it that thoroughly, but it's definitely possible, perhaps using typing.get_type_hints itself, or looking into what it does. I'm not sure what the best way is yet, but it is possible.

SplitInfinity commented:
Very cool - I will look into that and see if I can add objects referred to by type annotations to the callback constructed by `createResolutionCallbackForClassMethods`.

SplitInfinity commented:
Okay, so I think I have a plan but I wanted to run it by everyone first before I do it. `typing.get_type_hints(fn)` returns a `Dict[str, Any]` that maps argument name to the Python object representing the type it is annotated with. So, for the following function:
```
device_t = torch.device
def fn(x: int, y: device_t):
  pass
```
`typing.get_type_hints` returns `{'x': <class 'int'>, 'y': <class 'torch.device'>}`.

So what we could do is inspect the source code for each class method and insert into the internal map used by the resolution callback a mapping from whatever is used to annotate each argument in the source (interpreted as a string) to the corresponding type in the map returned by `typing.get_type_hints`. So for the function above, the resolution callback used for compiling the class would be augmented with `{"int": <class 'int'>, "device_t": <class 'torch.device'>}`.

Does that make sense?

Commit:
Update on "[JIT] Resolve `torch.device` in recursive compilation of classes"


**Summary**
This commit allows `torch.device` to be resolved properly when used in
class types that are recursively scripted. This is accomplished using
`string_to_type_lut`, similar to how `torch.Tensor` is resolved in this
situation.

Classes that are not explicitly annotated with `@torch.jit.script` are
implicitly scripted during the compilation of a function or class method
that uses them. One key difference between this method of class type
compilation and explicit scripting is that the former uses a resolution callback
that can only resolve variables that class methods close over (see
`_jit_internal.createResolutionCallbackForClassMethods`). This does
not include type annotations and default arguments. This means that
builtin types like `torch.Tensor` and `torch.device` cannot be resolved
using the resolution callback. This issue does not arise when explicitly
scripting classes because the resolution callback for that code path is
constructed from scope of the class definition
(see `_jit_internal.createResolutionCallbackFromFrame`). `torch.Tensor`
and `torch.device` are almost always present in that scope, usually from
`import`ing `torch`.

**Test Plan**
This commit adds a new unit test to `TestClassType`,
`test_recursive_script_builtin_type_resolution`.

**Fixes**
This commit closes #47405.

[ghstack-poisoned]

SplitInfinity commented:
I went ahead and did it anyway 👀 .

eellison commented on torch/_jit_internal.py:
```diff
@@ -238,6 +264,7 @@ def createResolutionCallbackForClassMethods(cls):
 
     for fn in fns:
         captures.update(get_closure(fn))
```
I think we're overr

eellison commented on torch/_jit_internal.py:
```diff
@@ -226,6 +226,32 @@ def can_compile_class(cls):
+    captures = {}
+    signature = inspect.signature(fn)
+
+    for name, parameter in signature.parameters.items():
```
does this also cover return annotations? maybe worth adding a test case.

SplitInfinity replied:
Let me add that. Also, I discovered that `str(signature.parameter['name'])` removes any aliases, so the type annotation that is displayed is the string representation of the type object, not the string used to annotate the parameter in the source code. So we will have to actually inspect the source and parse out the function declaration. I'm going to add that too and re-request review.

eellison reviewed (approved):
Cool, very clean!

Commit:
Update on "[JIT] Resolve `torch.device` in recursive compilation of classes"


**Summary**
This commit allows `torch.device` to be resolved properly when used in
class types that are recursively scripted. This is accomplished by augmenting
the resolution callback used during recursively class scripting to include
the type annotations used on class method declarations.

Classes that are not explicitly annotated with `@torch.jit.script` are
implicitly scripted during the compilation of a function or class method
that uses them. One key difference between this method of class type
compilation and explicit scripting is that the former uses a resolution callback
that can only resolve variables that class methods close over (see
`_jit_internal.createResolutionCallbackForClassMethods`). This does
not include type annotations and default arguments. This means that
builtin types like `torch.Tensor` and `torch.device` cannot be resolved
using the resolution callback. This issue does not arise when explicitly
scripting classes because the resolution callback for that code path is
constructed from scope of the class definition
(see `_jit_internal.createResolutionCallbackFromFrame`). `torch.Tensor`
and `torch.device` are almost always present in that scope, usually from
`import`ing `torch`.

**Test Plan**
This commit adds a new unit test to `TestClassType`,
`test_recursive_script_builtin_type_resolution`.

**Fixes**
This commit closes #47405.

[ghstack-poisoned]

Commit:
Update on "[JIT] Resolve `torch.device` in recursive compilation of classes"


**Summary**
This commit allows `torch.device` to be resolved properly when used in
class types that are recursively scripted. This is accomplished by augmenting
the resolution callback used during recursively class scripting to include
the type annotations used on class method declarations.

Classes that are not explicitly annotated with `@torch.jit.script` are
implicitly scripted during the compilation of a function or class method
that uses them. One key difference between this method of class type
compilation and explicit scripting is that the former uses a resolution callback
that can only resolve variables that class methods close over (see
`_jit_internal.createResolutionCallbackForClassMethods`). This does
not include type annotations and default arguments. This means that
builtin types like `torch.Tensor` and `torch.device` cannot be resolved
using the resolution callback. This issue does not arise when explicitly
scripting classes because the resolution callback for that code path is
constructed from scope of the class definition
(see `_jit_internal.createResolutionCallbackFromFrame`). `torch.Tensor`
and `torch.device` are almost always present in that scope, usually from
`import`ing `torch`.

**Test Plan**
This commit adds a new unit test to `TestClassType`,
`test_recursive_script_builtin_type_resolution`.

**Fixes**
This commit closes #47405.

[ghstack-poisoned]

Commit:
Update on "[JIT] Resolve `torch.device` in recursive compilation of classes"


**Summary**
This commit allows `torch.device` to be resolved properly when used in
class types that are recursively scripted. This is accomplished by augmenting
the resolution callback used during recursively class scripting to include
the type annotations used on class method declarations.

Classes that are not explicitly annotated with `@torch.jit.script` are
implicitly scripted during the compilation of a function or class method
that uses them. One key difference between this method of class type
compilation and explicit scripting is that the former uses a resolution callback
that can only resolve variables that class methods close over (see
`_jit_internal.createResolutionCallbackForClassMethods`). This does
not include type annotations and default arguments. This means that
builtin types like `torch.Tensor` and `torch.device` cannot be resolved
using the resolution callback. This issue does not arise when explicitly
scripting classes because the resolution callback for that code path is
constructed from scope of the class definition
(see `_jit_internal.createResolutionCallbackFromFrame`). `torch.Tensor`
and `torch.device` are almost always present in that scope, usually from
`import`ing `torch`.

**Test Plan**
This commit adds a new unit test to `TestClassType`,
`test_recursive_script_builtin_type_resolution`.

**Fixes**
This commit closes #47405.

[ghstack-poisoned]

SplitInfinity commented:
It turns out that `str(signature.parameter['name'])` removes any aliases, so the type annotation that is displayed is the string representation of the type object, not the string used to annotate the parameter in the source code. So we will have to actually inspect the source and parse out the function declaration. I can't use `frontend.py` since it imports `_jit_internal`, so I opted to use the `ast` module to parse the function source code and reconstruct the literal annotations.

Would it be better to parse the function declaration using the `re` module?

ppwwyyxx commented:
When I applied this PR it seems to break this use case:
```python
class Boxes:
    def __init__(self, tensor: torch.Tensor):
        self.tensor = tensor
    def clone(self) -> "Boxes":
        return Boxes(self.tensor.clone())

def forward(x: int):
    meta = Boxes(torch.tensor([1, 2, 3, 4]))
    new_obj = meta.clone()
    return new_obj.tensor + x

if __name__ == "__main__":
    m1 = torch.jit.script(forward)
```
gives:
```
  File "torch/_jit_internal.py", line 271, in get_annotation_str
    raise RuntimeError(f"Unexpected node type: {type(annotation)}")
RuntimeError: Unexpected node type: <class '_ast.Str'>
```


It works after adding `@torch.jit.script`. Perhaps this is why the original test (https://github.com/pytorch/pytorch/pull/39821/files) didn't catch it.

ppwwyyxx commented on torch/_jit_internal.py:
```diff
@@ -226,6 +228,68 @@ def can_compile_class(cls):
+        elif isinstance(annotation, ast.Constant) or isinstance(annotation, ast.NameConstant):
+            return f"{annotation.value}"
+
+        raise RuntimeError(f"Unexpected node type: {type(annotation)}")
```
Could you print `fn` as well? When I first encountered the error `RuntimeError: Unexpected node type: <class '_ast.Str'>
` (discussed in another comment) I have no clue where it comes from in a large model.

SplitInfinity replied:
Updated the diff to return `None` for unsupported nodes instead of throwing this error.

Commit:
Update on "[JIT] Resolve `torch.device` in recursive compilation of classes"


**Summary**
This commit allows `torch.device` to be resolved properly when used in
class types that are recursively scripted. This is accomplished by augmenting
the resolution callback used during recursively class scripting to include
the type annotations used on class method declarations.

Classes that are not explicitly annotated with `@torch.jit.script` are
implicitly scripted during the compilation of a function or class method
that uses them. One key difference between this method of class type
compilation and explicit scripting is that the former uses a resolution callback
that can only resolve variables that class methods close over (see
`_jit_internal.createResolutionCallbackForClassMethods`). This does
not include type annotations and default arguments. This means that
builtin types like `torch.Tensor` and `torch.device` cannot be resolved
using the resolution callback. This issue does not arise when explicitly
scripting classes because the resolution callback for that code path is
constructed from scope of the class definition
(see `_jit_internal.createResolutionCallbackFromFrame`). `torch.Tensor`
and `torch.device` are almost always present in that scope, usually from
`import`ing `torch`.

**Test Plan**
This commit adds a new unit test to `TestClassType`,
`test_recursive_script_builtin_type_resolution`.

**Fixes**
This commit closes #47405.

[ghstack-poisoned]

ppwwyyxx commented:
Thanks! I just tried the latest version and it worked well for d2

eellison commented on test/jit/test_class_type.py:
```diff
@@ -1308,3 +1308,36 @@ def fn() -> bool:
+            def h(self, x: device_t) -> device_t:
+                return x
+
+            def i(self) -> 'A':
```
Can you add a test case where `A` is an argument as well?

eellison commented on torch/_jit_internal.py:
```diff
@@ -226,6 +228,85 @@ def can_compile_class(cls):
+        if arg.annotation and get_annotation_str(arg.annotation) is not None
+    }
+
+    # Gather a dictionary of parameter name -> type, skipping any parameters whose annotated
```
IMO it might be easier to follow the logic of how we process an argument with a for loop, over processing each element, instead of the two `annotation_to_type` and `name_to_annotation` dictionaries. it also makes it easier to intersperse comments to its specific line of code

```
annotation_to_type = {}
for arg in f.args.args.:
     annotation_str = get_annotation_str(arg.annotation)
     if annotation_str is None:
          continue
    arg_name = arg.arg
    annotation_to_type[annotation] = name_to_type[arg_name]
```

eellison commented on test/jit/test_class_type.py:
```diff
@@ -1308,3 +1308,36 @@ def fn() -> bool:
+            def i(self) -> 'A':
+                return A()
+
+            def j(self, a: List[int]):
```
Could you add a test case where a return type has a contained type, like `List[device_t]` ? if that doesnt work it can be filed as a separate issue as well, this is already a pretty big PR

eellison reviewed (approved):
Looks great, well done!! I liked all of the comments for failure cases, I would have had no idea what was going on otherwise. Comments on the code are just IMO, feel free to accept some/none of them.

SplitInfinity commented on torch/_jit_internal.py:
```diff
@@ -226,6 +228,85 @@ def can_compile_class(cls):
+    # to annotate the arguments (e.g device_t = torch.device, and then d: device_t).
+    src = inspect.getsource(fn)
+
+    # frontend.py cannot be used here because it includes _jit_internal, so use ast instead.
```
`get_jit_def` could be reused, but that uses `build_def`, which uses several other functions in that file.

Commit:
Update on "[JIT] Resolve `torch.device` in recursive compilation of classes"


**Summary**
This commit allows `torch.device` to be resolved properly when used in
class types that are recursively scripted. This is accomplished by augmenting
the resolution callback used during recursively class scripting to include
the type annotations used on class method declarations.

Classes that are not explicitly annotated with `@torch.jit.script` are
implicitly scripted during the compilation of a function or class method
that uses them. One key difference between this method of class type
compilation and explicit scripting is that the former uses a resolution callback
that can only resolve variables that class methods close over (see
`_jit_internal.createResolutionCallbackForClassMethods`). This does
not include type annotations and default arguments. This means that
builtin types like `torch.Tensor` and `torch.device` cannot be resolved
using the resolution callback. This issue does not arise when explicitly
scripting classes because the resolution callback for that code path is
constructed from scope of the class definition
(see `_jit_internal.createResolutionCallbackFromFrame`). `torch.Tensor`
and `torch.device` are almost always present in that scope, usually from
`import`ing `torch`.

**Test Plan**
This commit adds a new unit test to `TestClassType`,
`test_recursive_script_builtin_type_resolution`.

**Fixes**
This commit closes #47405.

Differential Revision: [D24995374](https://our.internmc.facebook.com/intern/diff/D24995374)

[ghstack-poisoned]

Commit:
Update on "[JIT] Resolve `torch.device` in recursive compilation of classes"


**Summary**
This commit allows `torch.device` to be resolved properly when used in
class types that are recursively scripted. This is accomplished by augmenting
the resolution callback used during recursively class scripting to include
the type annotations used on class method declarations.

Classes that are not explicitly annotated with `@torch.jit.script` are
implicitly scripted during the compilation of a function or class method
that uses them. One key difference between this method of class type
compilation and explicit scripting is that the former uses a resolution callback
that can only resolve variables that class methods close over (see
`_jit_internal.createResolutionCallbackForClassMethods`). This does
not include type annotations and default arguments. This means that
builtin types like `torch.Tensor` and `torch.device` cannot be resolved
using the resolution callback. This issue does not arise when explicitly
scripting classes because the resolution callback for that code path is
constructed from scope of the class definition
(see `_jit_internal.createResolutionCallbackFromFrame`). `torch.Tensor`
and `torch.device` are almost always present in that scope, usually from
`import`ing `torch`.

**Test Plan**
This commit adds a new unit test to `TestClassType`,
`test_recursive_script_builtin_type_resolution`.

**Fixes**
This commit closes #47405.

Differential Revision: [D24995374](https://our.internmc.facebook.com/intern/diff/D24995374)

[ghstack-poisoned]

Commit:
Update on "[JIT] Resolve `torch.device` in recursive compilation of classes"


**Summary**
This commit allows `torch.device` to be resolved properly when used in
class types that are recursively scripted. This is accomplished by augmenting
the resolution callback used during recursively class scripting to include
the type annotations used on class method declarations.

Classes that are not explicitly annotated with `@torch.jit.script` are
implicitly scripted during the compilation of a function or class method
that uses them. One key difference between this method of class type
compilation and explicit scripting is that the former uses a resolution callback
that can only resolve variables that class methods close over (see
`_jit_internal.createResolutionCallbackForClassMethods`). This does
not include type annotations and default arguments. This means that
builtin types like `torch.Tensor` and `torch.device` cannot be resolved
using the resolution callback. This issue does not arise when explicitly
scripting classes because the resolution callback for that code path is
constructed from scope of the class definition
(see `_jit_internal.createResolutionCallbackFromFrame`). `torch.Tensor`
and `torch.device` are almost always present in that scope, usually from
`import`ing `torch`.

**Test Plan**
This commit adds a new unit test to `TestClassType`,
`test_recursive_script_builtin_type_resolution`.

**Fixes**
This commit closes #47405.

Differential Revision: [D24995374](https://our.internmc.facebook.com/intern/diff/D24995374)

[ghstack-poisoned]
